core/src/com/unciv/logic/map/UnitMovementAlgorithms.kt

```diff
@@ -121,52 +124,51 @@ class UnitMovementAlgorithms(val unit:MapUnit) {
+            tilesToCheck = newTilesToCheck.filterNot { tile -> tile.neighbors.all { it in newTilesToCheck || it in tilesToCheck } }
 
             distance++
         }
     }
 
-    fun getTileToMoveToThisTurn(finalDestination:TileInfo): TileInfo {
+    fun getTileToMoveToThisTurn(finalDestination: TileInfo): TileInfo {
 
         val currentTile = unit.getTile()
         if (currentTile == finalDestination) return currentTile
 
-        if(unit.type.isAirUnit()){
-            return finalDestination // head there directly
-        }
-
+        // head there directly
+        if (unit.type.isAirUnit()) return finalDestination
 
         val distanceToTiles = getDistanceToTiles()
 
-        if (distanceToTiles.containsKey(finalDestination)) { // we should be able to get there this turn
-            if (canMoveTo(finalDestination))
-                return finalDestination
+        class UnreachableDestinationException : Exception()
 
-            // Someone is blocking to the path to the final tile...
-            val destinationNeighbors = finalDestination.neighbors
-            if (destinationNeighbors.contains(currentTile)) // We're right nearby anyway, no need to move
-                return currentTile
+        // If the tile is far away, we need to build a path how to get there, and then take the first step
+        if (!distanceToTiles.containsKey(finalDestination))
+            return getShortestPath(finalDestination).firstOrNull()
+                    ?: throw UnreachableDestinationException()
 
-            val reachableDestinationNeighbors = destinationNeighbors
-                    .filter { distanceToTiles.containsKey(it) && canMoveTo(it) }
-            if (reachableDestinationNeighbors.isEmpty()) // We can't get closer...
-                return currentTile
+        // we should be able to get there this turn
+        if (canMoveTo(finalDestination))
+            return finalDestination
 
-            return reachableDestinationNeighbors.minBy { distanceToTiles[it]!!.totalDistance }!! // we can get a little closer
-        }   // If the tile is far away, we need to build a path how to get there, and then take the first step
-        val path = getShortestPath(finalDestination)
-        class UnreachableDestinationException : Exception()
-        if (path.isEmpty()) throw UnreachableDestinationException()
-        return path.first()
+        // Someone is blocking to the path to the final tile...
+        val destinationNeighbors = finalDestination.neighbors
+        return when (currentTile) {
+            in destinationNeighbors -> currentTile // We're right nearby anyway, no need to move
+            else -> destinationNeighbors.asSequence()
+                    .filter { distanceToTiles.containsKey(it) && canMoveTo(it) }
+                    .minBy { distanceToTiles.getValue(it).totalDistance } // we can get a little closer
+                    ?: currentTile // We can't get closer...
```

Comment:
slightly refactored here and uses sequences and minBy to check for isEmpty which will return currentTile if empty/null